modepy/nodes.py

```diff
@@ -291,31 +324,157 @@ def warp_and_blend_nodes(dims, n, node_tuples=None):
+    return tensor_product_nodes(dims, legendre_gauss_lobatto_nodes(n))
 
 # }}}
 
 
+# {{{ space-based interface
+
+@singledispatch
+def node_tuples_for_space(space: FunctionSpace) -> List[Tuple[int]]:
+    raise NotImplementedError(type(space).__name__)
+
+
+@singledispatch
+def equispaced_nodes_for_space(space: FunctionSpace, shape: Shape):
+    raise NotImplementedError((type(space).__name__, type(shape).__name))
+
+
+@singledispatch
+def edge_clustered_nodes_for_space(space: FunctionSpace, shape: Shape):
+    raise NotImplementedError((type(space).__name__, type(shape).__name))
+
+
+@singledispatch
+def random_nodes_for_shape(shape: Shape, nnodes: int, rng=None):
+    """
+    :arg generator: a :class:`numpy.random.Generator`.
+    :returns: a :class:`numpy.ndarray` that returns an array of
+        shape `(dim, nnodes)` of random nodes in the reference element.
+    """
+    raise NotImplementedError(type(shape).__name__)
+
+# }}}
+
+
+# {{{ PN
+
+@node_tuples_for_space.register(PN)
+def _(space: PN):
+    from pytools import \
+            generate_nonnegative_integer_tuples_summing_to_at_most as gnitsam
+    return tuple(gnitsam(space.order, space.spatial_dim))
+
+
+@equispaced_nodes_for_space.register(PN)
+def _(space: PN, shape: Simplex):
+    if not isinstance(shape, Simplex):
+        raise NotImplementedError((type(space).__name__, type(shape).__name))
+    if space.spatial_dim != shape.dim:
+        raise ValueError("spatial dimensions of shape and space must match")
+
+    return (np.array(node_tuples_for_space(space), dtype=np.float64)
+            / space.order*2 - 1).T
+
+
+@edge_clustered_nodes_for_space.register(PN)
+def _(space: PN, shape: Simplex):
+    if not isinstance(shape, Simplex):
+        raise NotImplementedError((type(space).__name__, type(shape).__name))
+    if space.spatial_dim != shape.dim:
+        raise ValueError("spatial dimensions of shape and space must match")
+
+    if space.spatial_dim <= 3:
+        return warp_and_blend_nodes(space.spatial_dim, space.order)
+    else:
+        return equidistant_nodes(space.spatial_dim, space.order)
+
+
+@random_nodes_for_shape.register(Simplex)
+def _(shape: Simplex, nnodes: int, rng=None):
+    if rng is None:
+        rng = np.random
+
+    result = np.zeros((shape.dim, nnodes))
+    nnodes_obtained = 0
+    while True:
+        new_nodes = rng.uniform(-1.0, 1.0, size=(shape.dim, nnodes-nnodes_obtained))
+        new_nodes = new_nodes[:, new_nodes.sum(axis=0) < 2-shape.dim]
+        nnew_nodes = new_nodes.shape[1]
+        result[:, nnodes_obtained:nnodes_obtained+nnew_nodes] = new_nodes
+        nnodes_obtained += nnew_nodes
+
+        if nnodes_obtained == nnodes:
+            return result
+
+# }}}
+
+
+# {{{ QN
+
+@node_tuples_for_space.register(QN)
+def _(space: QN):
+    from pytools import \
+            generate_nonnegative_integer_tuples_below as gnitb
+    return tuple([
+        tp[::-1] for tp in gnitb(space.order + 1, space.spatial_dim)
+        ])
```

Comment:
Reopening https://github.com/alexfikl/modepy/pull/1#discussion_r533931559, since I realized it might be a bit of a nasty hack.

I basically just transposed the node tuples so that getting equidistant nodes is the same in both of those cases.

Reply:
By "transposed", you mean the order reversal of the tuple? If so, then I think that's fine. I'd like to standardize on "reversed `gnitb` order" everywhere, to minimize the potential for confusion, and avoid the flips we battled in https://github.com/inducer/meshmode/pull/70. Does that sound reasonable?

Reply:
FWIW, the code here looks OK to me.

Reply:
Yes, flipped, not transposed!

Sounds good to me! Just wanted to make sure.